### PR TITLE
Add room sections

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

**Comment by KrisXV** [comment]
Client-side portion of https://github.com/smogon/pokemon-showdown/pull/8205

File: js/client-rooms.js [hunk 1]
```diff
@@ -87,9 +87,39 @@
...
+				(rooms.pspl && rooms.pspl.length ? '<h2 class="rooms-psplchatrooms">PSPL Winner</h2>' + _.map(rooms.pspl.sort(this.compareRooms), this.renderRoomBtn).join("") : '')
+			);
+			var buf = '';
+			var hardcodedSections = ['officialtiers', 'communityprojects', 'languages'];
+			for (var i = 0; i < hardcodedSections.length; i++) {
+				var section = sections[hardcodedSections[i]];
+				if (!section) continue;
+				/*section = section.filter(function (x) {
+					return (rooms.pspl || []).indexOf(x) < 0;
+				});*/
```
Comment (AnnikaCodes) [inline js/client-rooms.js:102]: ?

File: js/client-rooms.js [hunk 2]
```diff
@@ -87,9 +87,39 @@
...
+				if (!section.length) continue;
+				buf += '<h2 class="rooms-chatrooms">' + ((rooms.sectionTitles || {})[hardcodedSections[i]] || hardcodedSections[i]) + '</h2>' + _.map(section.sort(this.compareRooms), this.renderRoomBtn).join("");
+			}
+			for (var j = 0; j < Object.keys(sections).sort().length; j++) {
```
Comment (AnnikaCodes) [inline js/client-rooms.js:106]:
```suggestion
			for (var j = 0; j < Object.keys(sections).length; j++) {
```
Sorting an array won't change its length

Comment (KrisXV) [reply js/client-rooms.js:106]: It’s sorted per aeonics request to alphabetize the rest of the categories

File: js/client-rooms.js (hunk 1 above)
Comment (KrisXV) [reply js/client-rooms.js:102]: Idk why it’s commented I’ll undo when I get home

File: js/client-rooms.js (hunk 2 above)
Comment (AnnikaCodes) [reply js/client-rooms.js:106]: You sort it in the body of the loop; here, you're just getting the length of the array, which sorting doesn't affect.

**Comment by KrisXV** [comment]
@Zarel can you please review this?

https://i.imgur.com/pXMdkem.gif heres the gif i posted in the below log (cant embed, it's too large of a file)

> [10:20 PM] Kris: @Zarel please review the client code for room sections
[10:21 PM] Kris: I made a select list to filter by section and posted a gif in here (which I also tagged you in), ideally this goes live 24h from now so a review would be much appreciated
[10:22 PM] Aeo: Confirming the above. This is pretty urgent now.
[10:23 PM] Kris: Annika and Mia already reviewed server side viciously because I wanted at least 2 other pairs of eyes on it

File: js/client-rooms.js [hunk 3]
```diff
@@ -87,9 +101,39 @@
...
+			var officialRooms = allRooms.filter(function (roomData) {
+				return roomData.section === 'Official';
+			});
+			if (officialRooms.length) {
+				allRooms = allRooms.filter(function (roomData) {
+					return roomData.section !== 'Official';
+				});
+			}
+			this.$('.roomlist').first().html(
+				(officialRooms.length ? '<h2 class="rooms-officialchatrooms">Official chat rooms</h2>' + _.map(officialRooms.sort(this.compareRooms), this.renderRoomBtn).join("") : '') +
```
Comment (Zarel) [inline js/client-rooms.js:131]: tbh just add `map` to the polyfill...

Comment (Zarel) [reply js/client-rooms.js:131]: Ah, this is old code, probably best deferred for later.

File: js/client-rooms.js [hunk 4]
```diff
@@ -28,7 +40,7 @@
 				this.lastUpdate = new Date().getTime();
 			}
 			var prevPos = this.$el.scrollTop();
-			this.$('button[name=joinRoomPopup]').focus();
+			// this.$('button[name=joinRoomPopup]').focus();
```
Comment (Zarel) [inline js/client-rooms.js:43]: Why?

Comment (Zarel) [reply js/client-rooms.js:43]: This is important for keyboard shortcuts, too.

Comment (KrisXV) [reply js/client-rooms.js:43]: Constantly re-focusing this button makes it impossible to use the dropdown

File: js/client-rooms.js [hunk 5]
```diff
@@ -87,9 +101,39 @@
 				var rightSide = '<button class="button" name="roomlist" title="Watch an active battle"><span class="pixelated battlecount" title="Meloetta is PS\'s mascot! The Pirouette forme is Fighting-type, and represents our battles." ></span><strong>' + battleCount + '</strong> active ' + (battleCount == 1 ? 'battle' : 'battles') + '</button>';
 				this.$('.roomlisttop').html('<div class="roomcounters">' + leftSide + '</td><td>' + rightSide + '</div>');
 			}
-			this.$('.roomlist').first().html('<h2 class="rooms-officialchatrooms">Official chat rooms</h2>' + _.map(rooms.official, this.renderRoomBtn).join("") +
-				(rooms.pspl && rooms.pspl.length ? '<a href="https://www.smogon.com/forums/threads/3649563/" target="_blank"><h2 class="rooms-psplchatrooms">PSPL Winner</h2></a>' + _.map(rooms.pspl, this.renderRoomBtn).join("") : ''));
-			this.$('.roomlist').last().html('<h2 class="rooms-chatrooms">Chat rooms</h2>' + _.map(rooms.chat.sort(this.compareRooms), this.renderRoomBtn).join(""));
+			var allRooms = rooms.chat;
+			var psplRooms = allRooms.filter(function (roomData) {
+				return roomData.spotlight === "PSPL Winner" && roomData.section !== 'Official';
+			});
```
Comment (Zarel) [inline js/client-rooms.js:107]: This does not seem to be compatible with the old protocol.

File: js/client-rooms.js [hunk 6]
```diff
@@ -19,6 +23,14 @@
 			app.user.on('change:named', this.updateUser, this);
 			this.update();
 		},
+		initSectionSelection: function () {
+			var buf = ['<option value="all">All rooms</option>'];
+			for (var i = 0; i < app.roomsData.sectionTitles.length; i++) {
+				var sectionName = app.roomsData.sectionTitles[i];
+				buf.push('<option value="' + toID(sectionName) + '">' + sectionName + '</option>');
```
Comment (Zarel) [inline js/client-rooms.js:30]: Say it with me: html~ injection~ vulnerability~

Comment (Zarel) [reply js/client-rooms.js:30]: (Say it with me: Finishing Preact is a better use of your time than making a bunch of UI changes to the legacy client)

File: js/client-rooms.js [hunk 7]
```diff
@@ -5,11 +5,17 @@
...
+			'change select[name=sections]': 'refresh'
+		},
 		isSideRoom: true,
 		initialize: function () {
+			this.focusedSection = '';
+
 			this.$el.addClass('ps-room-light').addClass('scrollable');
 			var buf = '<div class="pad"><button class="button" style="float:right;font-size:10pt;margin-top:3px" name="closeHide"><i class="fa fa-caret-right"></i> Hide</button>';
-			buf += '<div class="roomlisttop"></div><div class="roomlist"><p><em style="font-size:20pt">Loading...</em></p></div><div class="roomlist"></div>';
+			buf += '<div class="roomlisttop"></div><p>Rooms filter: <select name="sections"><option value="all">All rooms</option></select></p>';
```
Comment (KrisXV) [inline js/client-rooms.js:17]:
```suggestion
			buf += '<div class="roomlisttop"></div><p>Rooms filter: <select name="sections"><option value="all">(All rooms)</option></select></p>';
```

**Pull request merged**